File: src/main/java/org/example/tree/ValidBST.java
package org.example.tree;

public class ValidBST {

    public boolean isValidBST(TreeNode root) {
        return isValid(root, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
    }

    private boolean isValid(TreeNode node, double left, double right) {
        if (node == null) return true;
        if (!(node.val > left && node.val < right)) return false;

        return (isValid(node.left, left, node.val) && isValid(node.right, node.val, right));
    }
}

File: src/main/java/org/example/tree/ValidBST.py
from typing import Optional

from org.example.tree.TreeNode import TreeNode


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(node, minVal, maxVal):
            if not node: return True
            if not (minVal < node.val < maxVal):
                return False
            return (dfs(node.left, minVal, node.val) and
                    dfs(node.right, node.val, maxVal))

        return dfs(root, float("-inf"), float("inf"))


n1 = TreeNode(1)
n3 = TreeNode(3)
n2 = TreeNode(2, n1, n3)
obj = Solution()
print(obj.isValidBST(n2))
